integration/setup/listener.go

```diff
@@ -101,71 +83,157 @@ func setupListener(tb testtb.TB, ctx context.Context, logger *zap.Logger) string
 	defer span.End()
 
 	defer observability.FuncCall(ctx)()
+	var l *clientconn.Listener
+	var handler string
 
-	require.Empty(tb, *targetURLF, "-target-url must be empty for in-process FerretDB")
+	if *shareServerF {
+		l = listener
+		handler = handlerType
+	} else {
+		handler, l = initListener(ctx)
 
-	var handler string
+		runDone := make(chan struct{})
+
+		go func() {
+			defer close(runDone)
+
+			err := l.Run(ctx)
+			if err == nil || errors.Is(err, context.Canceled) {
+				logger.Info("Listener stopped without error")
+			} else {
+				logger.Error("Listener stopped", zap.Error(err))
+			}
+		}()
+
+		// ensure that all listener's and handler's logs are written before test ends
+		tb.Cleanup(func() {
+			<-runDone
+		})
+	}
+
+	var hostPort, unixSocketPath string
+	var tlsAndAuth bool
+
+	switch {
+	case *targetTLSF:
+		hostPort = l.TLSAddr().String()
+		tlsAndAuth = true
+	case *targetUnixSocketF:
+		unixSocketPath = l.UnixAddr().String()
+	default:
+		hostPort = l.TCPAddr().String()
+	}
+
+	uri := listenerMongoDBURI(tb, hostPort, unixSocketPath, tlsAndAuth)
+
+	logger.Info("Listener started", zap.String("handler", handler), zap.String("uri", uri))
+
+	return uri
+}
+
+func initListener(ctx context.Context) (string, *clientconn.Listener) {
```

Comment:
I think that function could accept `fatalf func(format string, args ...any)` that could be either `tb.Fatalf` or `zap.S().Fatalf`